text/0000-kadalu-thin-arbiter-support.md

```diff
@@ -0,0 +1,84 @@
+
+
+So, the thin-arbiter feature is very useful for users who have just 2 data centers with required latency, but not the 3rd. With this feature, all such users can actually have the 3rd tie-breaker node in cloud without latency limitation.
+
+Even for kadalu users, if you have a stretched k8s cluster, which spans across 2 data centers (or 2 availability zones in Cloud), this feature will be useful. This helps by reducing the resources required for high-availability, and also the speed of write (1/2 instead of 1/3).
+
+The design details of thin-arbiter was discussed [here](https://github.com/gluster/glusterfs/issues/352) and [here](https://docs.gluster.org/en/latest/Administrator%20Guide/Thin-Arbiter-Volumes/). In this proposal we will talk about how we can add thin-arbiter volume type to kadalu.
+
+
+## Design
+
+Kadalu works with predefined volume templates for gluster. We need add thin-arbiter template, and also define a specific storage `Type` for this template. The same can be used for storage-config 'type' too. Already `Replica1` and `Replica3` namespace taken. We can choose 'Replica2' as it is indeed a replica2 setup.
+
+Proposed config file looks something like below:
+
+```yaml
+# File: storage-config.yaml
+---
+apiVersion: kadalu-operator.storage/v1alpha1
+kind: KadaluStorage
+metadata:
+ # This will be used as name of PV Hosting Volume
+  name: storage-pool-name
+spec:
+  type: Replica2       # Just to highlight your data is just in 2 places.
+  storage:
+    - node: kube1      # node name as shown in `kubectl get nodes`
+      device: /dev/vdc # Device to provide storage to all PVs
+    - node: kube2      # node name as shown in `kubectl get nodes`
+      device: /dev/vdc # Device to provide storage to all PVs
+    - tie-breaker:
+        deployment: local # other option is 'external'.
+        node: tie-breaker.kadalu.io
+        path: /mnt/tie-breaker/
+```
+
+
+In the PV spec, one would need to use the option as below:
+
+```yaml
+# -*- mode: yaml -*-
+---
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: pv1
+spec:
+  storageClassName: kadalu.replica2
+  accessModes:
+    - ReadWriteOnce # You can also provide 'ReadWriteMany' here
+  resources:
+    requests:
+      storage: 1Gi
+```
+
+The main definition which needs to be properly validated would be of `tie-breaker`. If it has field which says `deployment: local`, then kadalu operator should deploy the thin-arbiter locally in the node. If it is `external` then, we can assume it is handled outside.
+
+
+## Further improvement
+
+In the similar lines, we can add support for 'Arbiter' too, as there may be users who would like to have arbiter support in kadalu. Need to have a reserved type for the same (may be 'Arbiter'). But, it doesn't make sense to expose that option much if thin-arbiter works fine for us.
+
```

Comment:
I think it would be better to have separate types (maybe Replica2-Arb or Replica2-TArb) for arbiter and thin-arbiter so that if user wants to use normal Replica2 knowing its limitations fully, we can support that template. I'm saying this because I still see occasional mails on gluster-users that use (distributed) replica 2 setups, so giving options would be nice.

Reply:
Makes sense. Need to think about a good name for config :)

Reply:
I am inclined to have 'Thin-Arbiter' as the default 'Replica2' type. We will still come up with different names for Arbiter and `replica 2` mode of gluster. I feel, in the soon to be seen future, this 'gluster' naming would be bit separate from kadalu config names. So, as we support other configs, we can have names figured out.

Reply:
Makes sense. We can add `without-tiebreaker: true` or something when needed.